File: update_liga.py
#!/usr/bin/env python3
"""
Pobiera tabelę i terminarz z 90minut.pl i zapisuje je do data/liga.json,
skąd czyta je strona klubu. Uruchamiany raz dziennie (launchd / cron).

Użycie:
    python3 update_liga.py            # normalny przebieg
    python3 update_liga.py --dry-run  # wypisz wynik, nie zapisuj pliku
"""

import json
import os
import re
import sys
import urllib.request
from datetime import datetime
from html import unescape

URL = "http://www.90minut.pl/liga/1/liga15094.html"
NASZ_KLUB = "Janovia Janowiec"

BASE = os.path.dirname(os.path.abspath(__file__))
OUT = os.path.join(BASE, "data", "liga.json")

MIESIACE = {
    "stycznia": 1, "lutego": 2, "marca": 3, "kwietnia": 4,
    "maja": 5, "czerwca": 6, "lipca": 7, "sierpnia": 8,
    "września": 9, "wrzesnia": 9, "października": 10, "pazdziernika": 10,
    "listopada": 11, "grudnia": 12,
}


def pobierz(url):
    req = urllib.request.Request(url, headers={
        "User-Agent": "KS Janovia Janowiec - aktualizacja tabeli (1x dziennie)",
    })
    with urllib.request.urlopen(req, timeout=30) as r:
        return r.read().decode("iso-8859-2", errors="replace")


def czysc(s):
    """Usuwa tagi, encje i nadmiarowe spacje z fragmentu HTML."""
    s = re.sub(r"<[^>]+>", "", s)
    return unescape(s).replace("\xa0", " ").strip()


def sezon_lata(html):
    """Z tytułu 'Klasa B 2026/2027, grupa: ...' wyciąga (2026, 2027)."""
    m = re.search(r"(\d{4})\s*/\s*(\d{4})", html)
    if m:
        return int(m.group(1)), int(m.group(2))
    rok = datetime.now().year
    return rok, rok + 1


def nazwa_ligi(html):
    m = re.search(r"<title>(.*?)</title>", html, re.S)
    return czysc(m.group(1)) if m else ""


def parsuj_tabele(html):
    """
    Wiersz tabeli: <td><b>poz</b></td><td align=left>klub</td>
    potem M, Pkt, Z, R, P, Bramki (kolumna RAZEM).
    """
    tabela = []
    wzor = re.compile(
        r"<tr[^>]*>\s*"
        r"<td><b>(\d*)\.?</b></td>\s*"
        r'<td align="left">(.*?)</td>\s*'
        r"<td>(\d+)</td>\s*"
        r"<td><b>(-?\d+)</b></td>\s*"
        r"<td>(\d+)</td>\s*<td>(\d+)</td>\s*<td>(\d+)</td>\s*"
        r"<td>(\d+\s*-\s*\d+)</td>",
        re.S,
    )

    ostatnia_pozycja = 0
    for m in wzor.finditer(html):
        poz = int(m.group(1)) if m.group(1) else ostatnia_pozycja + 1
        ostatnia_pozycja = poz
        tabela.append({
            "poz": poz,
            "klub": czysc(m.group(2)),
            "m": int(m.group(3)),
            "pkt": int(m.group(4)),
            "z": int(m.group(5)),
            "r": int(m.group(6)),
            "p": int(m.group(7)),
            "bramki": czysc(m.group(8)).replace(" ", ""),
        })

    return tabela


def parsuj_terminarz(html, rok_jesien, rok_wiosna):
    """
    Nagłówek kolejki:  <b><u>Kolejka 3 - 5-6 września</u></b>
    Mecz: 4 komórki — gospodarz | wynik lub '-' | gość | data i godzina
    """
    mecze = []
    kolejka = None

    naglowek = re.compile(r"<b><u>\s*Kolejka\s*(\d+)\s*-\s*([^<]*)</u></b>", re.S)
    wiersz = re.compile(
        r'<td nowrap valign="top" width="180">(.*?)</td>\s*'
        r'<td nowrap valign="top" align="center" width="50">(.*?)</td>\s*'
        r'<td nowrap valign="top" width="180">(.*?)</td>\s*'
        r'<td valign="top" nowrap align="left" width="190">(.*?)</td>',
        re.S,
    )

    znaczniki = [(m.start(), m) for m in naglowek.finditer(html)]

    for i, (start, m) in enumerate(znaczniki):
        kolejka = int(m.group(1))
        etykieta = czysc(m.group(2))
        koniec = znaczniki[i + 1][0] if i + 1 < len(znaczniki) else len(html)

        for w in wiersz.finditer(html, start, koniec):
            gospodarz = czysc(w.group(1))
            wynik = czysc(w.group(2))
            gosc = czysc(w.group(3))
            termin = czysc(w.group(4))

            if not gospodarz or not gosc:
                continue

            mecze.append({
                "kolejka": kolejka,
                "kolejka_opis": etykieta,
                "gospodarz": gospodarz,
                "gosc": gosc,
                "wynik": wynik if re.match(r"^\d+\s*-\s*\d+$", wynik) else None,
                "termin": termin or None,
                # dokładny termin bywa pusty — wtedy bierzemy pierwszy dzień z nagłówka kolejki
                "data_iso": (na_iso(termin, rok_jesien, rok_wiosna)
                             or na_iso(etykieta, rok_jesien, rok_wiosna)),
                "data_przyblizona": not bool(na_iso(termin, rok_jesien, rok_wiosna)),
            })

    return mecze


def na_iso(termin, rok_jesien, rok_wiosna):
    """'9 sierpnia, 16:00' -> '2026-08-09T16:00:00'. Bez godziny -> data o 00:00."""
    if not termin:
        return None

    # obsługuje '9 sierpnia, 16:00' oraz zakres z nagłówka kolejki '22-23 sierpnia'
    m = re.match(
        r"(\d{1,2})(?:\s*-\s*\d{1,2})?\s+([a-ząćęłńóśźż]+)(?:,\s*(\d{1,2}):(\d{2}))?",
        termin, re.I,
    )
    if not m:
        return None

    miesiac = MIESIACE.get(m.group(2).lower())
    if not miesiac:
        return None

    # sezon przełamuje się na Nowy Rok: lipiec–grudzień to pierwszy rok
    rok = rok_jesien if miesiac >= 7 else rok_wiosna

    try:
        return datetime(
            rok, miesiac, int(m.group(1)),
            int(m.group(3) or 0), int(m.group(4) or 0)
        ).isoformat()
    except ValueError:
        return None


def main():
    dry = "--dry-run" in sys.argv

    html = pobierz(URL)
    rok_jesien, rok_wiosna = sezon_lata(html)

    tabela = parsuj_tabele(html)
    mecze = parsuj_terminarz(html, rok_jesien, rok_wiosna)

    if not tabela and not mecze:
        raise SystemExit("Nie udało się nic sparsować — strona 90minut zmieniła układ?")

    nasze = [m for m in mecze if NASZ_KLUB in (m["gospodarz"], m["gosc"])]
    nadchodzace = [m for m in nasze if not m["wynik"]]
    rozegrane = [m for m in nasze if m["wynik"]]

    dane = {
        "zrodlo": URL,
        "liga": nazwa_ligi(html),
        "klub": NASZ_KLUB,
        "zaktualizowano": datetime.now().isoformat(timespec="seconds"),
        "tabela": tabela,
        "nadchodzace": nadchodzace[:8],
        "ostatni": rozegrane[-1] if rozegrane else None,
        "wszystkie_mecze": mecze,
    }

    tekst = json.dumps(dane, ensure_ascii=False, indent=2)

    if dry:
        print(tekst)
        return

    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    with open(OUT, "w", encoding="utf-8") as f:
        f.write(tekst + "\n")

    print("Zapisano {}: {} zespołów w tabeli, {} meczów, {} nadchodzących.".format(
        os.path.relpath(OUT, BASE), len(tabela), len(mecze), len(nadchodzace)
    ))


if __name__ == "__main__":
    main()

File: update_liga.php
<?php
/**
 * Pobiera tabelę i terminarz z 90minut.pl i zapisuje je do data/liga.json.
 *
 * Odpowiednik update_liga.py, przepisany na PHP, żeby dane odświeżał sam
 * serwer — hosting współdzielony daje PHP, ale nie Pythona.
 *
 * Uruchomienie z crona:      php /sciezka/do/update_liga.php
 * Podgląd bez zapisu:        php update_liga.php --dry-run
 *
 * Z panelu wołane przez aktualizuj_lige(); wtedy plik jest tylko dołączany
 * i nic sam z siebie nie robi — patrz warunek na końcu.
 */

declare(strict_types=1);

const LIGA_URL       = 'http://www.90minut.pl/liga/1/liga15094.html';
const LIGA_NASZ_KLUB = 'Janovia Janowiec';

const LIGA_MIESIACE = [
    'stycznia' => 1, 'lutego' => 2, 'marca' => 3, 'kwietnia' => 4,
    'maja' => 5, 'czerwca' => 6, 'lipca' => 7, 'sierpnia' => 8,
    'września' => 9, 'wrzesnia' => 9, 'października' => 10, 'pazdziernika' => 10,
    'listopada' => 11, 'grudnia' => 12,
];

/**
 * 90minut serwuje stronę w ISO-8859-2, więc bez przekodowania polskie znaki
 * w nazwach klubów rozsypią się na krzaki.
 */
function liga_pobierz(string $url): string
{
    $naglowek = 'KS Janovia Janowiec - aktualizacja tabeli (1x dziennie)';

    if (function_exists('curl_init')) {
        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_TIMEOUT        => 30,
            CURLOPT_USERAGENT      => $naglowek,
        ]);
        $tresc = curl_exec($ch);
        $blad  = curl_error($ch);
        // bez curl_close() — od PHP 8.0 nic nie robi, a od 8.5 wypisuje
        // ostrzeżenie, które psułoby JSON wypisywany przy --dry-run

        if ($tresc === false) {
            throw new RuntimeException('Nie udało się pobrać strony: ' . $blad);
        }
    } else {
        $kontekst = stream_context_create(['http' => [
            'timeout' => 30,
            'header'  => 'User-Agent: ' . $naglowek,
        ]]);
        $tresc = @file_get_contents($url, false, $kontekst);

        if ($tresc === false) {
            throw new RuntimeException('Nie udało się pobrać strony (brak cURL i allow_url_fopen?).');
        }
    }

    return mb_convert_encoding($tresc, 'UTF-8', 'ISO-8859-2');
}

/** Usuwa tagi, encje i nadmiarowe spacje z fragmentu HTML. */
function liga_czysc(string $s): string
{
    $s = preg_replace('/<[^>]+>/', '', $s);
    $s = html_entity_decode($s, ENT_QUOTES | ENT_HTML5, 'UTF-8');

    // twarda spacja z encji &nbsp; to U+00A0, nie zwykły odstęp
    return trim(str_replace("\xC2\xA0", ' ', $s));
}

/** Z tytułu „Klasa B 2026/2027, grupa: …" wyciąga lata sezonu. */
function liga_sezon_lata(string $html): array
{
    if (preg_match('~(\d{4})\s*/\s*(\d{4})~', $html, $m)) {
        return [(int) $m[1], (int) $m[2]];
    }

    $rok = (int) date('Y');
    return [$rok, $rok + 1];
}

function liga_nazwa(string $html): string
{
    return preg_match('~<title>(.*?)</title>~s', $html, $m) ? liga_czysc($m[1]) : '';
}

/**
 * Wiersz tabeli: pozycja, klub, M, Pkt, Z, R, P, bramki.
 * Pozycja bywa pusta przy klubach ex aequo — wtedy liczymy dalej od poprzedniej.
 */
function liga_parsuj_tabele(string $html): array
{
    $wzor = '~<tr[^>]*>\s*'
          . '<td><b>(\d*)\.?</b></td>\s*'
          . '<td align="left">(.*?)</td>\s*'
          . '<td>(\d+)</td>\s*'
          . '<td><b>(-?\d+)</b></td>\s*'
          . '<td>(\d+)</td>\s*<td>(\d+)</td>\s*<td>(\d+)</td>\s*'
          . '<td>(\d+\s*-\s*\d+)</td>~s';

    $tabela = [];
    $ostatnia = 0;

    if (preg_match_all($wzor, $html, $trafienia, PREG_SET_ORDER)) {
        foreach ($trafienia as $m) {
            $poz = $m[1] !== '' ? (int) $m[1] : $ostatnia + 1;
            $ostatnia = $poz;

            $tabela[] = [
                'poz'    => $poz,
                'klub'   => liga_czysc($m[2]),
                'm'      => (int) $m[3],
                'pkt'    => (int) $m[4],
                'z'      => (int) $m[5],
                'r'      => (int) $m[6],
                'p'      => (int) $m[7],
                'bramki' => str_replace(' ', '', liga_czysc($m[8])),
            ];
        }
    }

    return $tabela;
}

/**
 * „9 sierpnia, 16:00" -> „2026-08-09T16:00:00". Bez godziny -> data o 00:00.
 * Obsługuje też zakres z nagłówka kolejki („22-23 sierpnia").
 */
function liga_na_iso(?string $termin, int $rokJesien, int $rokWiosna): ?string
{
    if (!$termin) { return null; }

    $wzor = '~^(\d{1,2})(?:\s*-\s*\d{1,2})?\s+([a-ząćęłńóśźż]+)(?:,\s*(\d{1,2}):(\d{2}))?~iu';

    if (!preg_match($wzor, $termin, $m)) { return null; }

    $miesiac = LIGA_MIESIACE[mb_strtolower($m[2], 'UTF-8')] ?? null;
    if (!$miesiac) { return null; }

    // sezon przełamuje się na Nowy Rok: lipiec–grudzień to pierwszy rok
    $rok = $miesiac >= 7 ? $rokJesien : $rokWiosna;

    $dzien   = (int) $m[1];
    $godzina = isset($m[3]) ? (int) $m[3] : 0;
    $minuta  = isset($m[4]) ? (int) $m[4] : 0;

    if (!checkdate($miesiac, $dzien, $rok)) { return null; }

    return sprintf('%04d-%02d-%02dT%02d:%02d:00', $rok, $miesiac, $dzien, $godzina, $minuta);
}

/**
 * Nagłówek kolejki: <b><u>Kolejka 3 - 5-6 września</u></b>
 * Mecz: cztery komórki — gospodarz | wynik lub „-" | gość | termin.
 *
 * Data w nagłówku bywa pominięta (runda wiosenna bez ustalonych terminów),
 * dlatego jest opcjonalna — inaczej mecze wpadałyby do poprzedniej kolejki.
 */
function liga_parsuj_terminarz(string $html, int $rokJesien, int $rokWiosna): array
{
    $wzorNaglowka = '~<b><u>\s*Kolejka\s*(\d+)\s*(?:-\s*([^<]*?))?\s*</u></b>~s';
    $wzorWiersza  = '~<td nowrap valign="top" width="180">(.*?)</td>\s*'
                  . '<td nowrap valign="top" align="center" width="50">(.*?)</td>\s*'
                  . '<td nowrap valign="top" width="180">(.*?)</td>\s*'
                  . '<td valign="top" nowrap align="left" width="190">(.*?)</td>~s';

    preg_match_all($wzorNaglowka, $html, $naglowki, PREG_SET_ORDER | PREG_OFFSET_CAPTURE);

    $mecze = [];
    $ile = count($naglowki);

    for ($i = 0; $i < $ile; $i++) {
        $kolejka  = (int) $naglowki[$i][1][0];
        $etykieta = liga_czysc($naglowki[$i][2][0] ?? '');

        $start  = $naglowki[$i][0][1];
        $koniec = ($i + 1 < $ile) ? $naglowki[$i + 1][0][1] : strlen($html);
        $blok   = substr($html, $start, $koniec - $start);

        if (!preg_match_all($wzorWiersza, $blok, $wiersze, PREG_SET_ORDER)) {
            continue;
        }

        foreach ($wiersze as $w) {
            $gospodarz = liga_czysc($w[1]);
            $gosc      = liga_czysc($w[3]);

            if ($gospodarz === '' || $gosc === '') { continue; }

            $wynik  = liga_czysc($w[2]);
            $termin = liga_czysc($w[4]);
            $zTerminu = liga_na_iso($termin, $rokJesien, $rokWiosna);

            $mecze[] = [
                'kolejka'          => $kolejka,
                'kolejka_opis'     => $etykieta,
                'gospodarz'        => $gospodarz,
                'gosc'             => $gosc,
                'wynik'            => preg_match('~^\d+\s*-\s*\d+$~', $wynik) ? $wynik : null,
                'termin'           => $termin !== '' ? $termin : null,
                // dokładny termin bywa pusty — wtedy bierzemy dzień z nagłówka kolejki
                'data_iso'         => $zTerminu ?? liga_na_iso($etykieta, $rokJesien, $rokWiosna),
                'data_przyblizona' => $zTerminu === null,
            ];
        }
    }

    return $mecze;
}

/**
 * Komplet spotkań ostatniej kolejki, w której padł jakikolwiek wynik.
 * Zanim sezon ruszy — najbliższa kolejka z samymi terminami.
 */
function liga_ostatnia_kolejka(array $mecze): ?array
{
    if (!$mecze) { return null; }

    $rozegrane = array_filter($mecze, static fn(array $m): bool => (bool) $m['wynik']);

    $numery = array_column($rozegrane ?: $mecze, 'kolejka');
    $numer  = $rozegrane ? max($numery) : min($numery);

    $wybrane = array_values(array_filter(
        $mecze,
        static fn(array $m): bool => $m['kolejka'] === $numer
    ));

    return [
        'numer'     => $numer,
        'opis'      => $wybrane ? $wybrane[0]['kolejka_opis'] : '',
        'rozegrana' => (bool) array_filter($wybrane, static fn(array $m): bool => (bool) $m['wynik']),
        'mecze'     => $wybrane,
    ];
}

/** Dla każdego klubu ostatnie `ile` meczów: Z (wygrana), R (remis), P (porażka). */
function liga_policz_forme(array $mecze, int $ile = 5): array
{
    $rozegrane = array_values(array_filter($mecze, static fn(array $m): bool => (bool) $m['wynik']));
    usort($rozegrane, static fn(array $a, array $b): int => $a['kolejka'] <=> $b['kolejka']);

    $forma = [];

    foreach ($rozegrane as $m) {
        $czesci = preg_split('~\s*-\s*~', $m['wynik']);
        if (count($czesci) < 2 || !is_numeric($czesci[0]) || !is_numeric($czesci[1])) {
            continue;
        }

        $golGosp = (int) $czesci[0];
        $golGosc = (int) $czesci[1];

        $strony = [
            [$m['gospodarz'], $golGosp, $golGosc, $m['gosc'], 'u siebie'],
            [$m['gosc'], $golGosc, $golGosp, $m['gospodarz'], 'na wyjeździe'],
        ];

        foreach ($strony as [$klub, $swoje, $obce, $rywal, $gdzie]) {
            $znak = $swoje > $obce ? 'Z' : ($swoje < $obce ? 'P' : 'R');

            $forma[$klub][] = [
                'wynik'    => $znak,
                'rywal'    => $rywal,
                'rezultat' => $swoje . '-' . $obce,
                'kolejka'  => $m['kolejka'],
                'gdzie'    => $gdzie,
            ];
        }
    }

    foreach ($forma as $klub => $lista) {
        $forma[$klub] = array_slice($lista, -$ile);
    }

    return $forma;
}

/**
 * Python zapisywał JSON z wcięciem dwóch spacji, PHP wcina czterema.
 * Zrównujemy, żeby przejście na PHP nie wygenerowało różnicy w każdej linii.
 */
function liga_json(array $dane): string
{
    $json = json_encode($dane, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_PRETTY_PRINT);

    return preg_replace_callback(
        '/^(?: {4})+/m',
        static fn(array $m): string => str_repeat(' ', strlen($m[0]) / 2),
        $json
    );
}

/**
 * Pobiera, parsuje i zapisuje. Zwraca podsumowanie dla panelu i crona.
 */
function aktualizuj_lige(bool $dryRun = false): array
{
    $html = liga_pobierz(LIGA_URL);
    [$rokJesien, $rokWiosna] = liga_sezon_lata($html);

    $tabela = liga_parsuj_tabele($html);
    $mecze  = liga_parsuj_terminarz($html, $rokJesien, $rokWiosna);

    if (!$tabela && !$mecze) {
        throw new RuntimeException('Nie udało się nic sparsować — 90minut zmieniło układ strony?');
    }

    $nasze = array_values(array_filter($mecze, static fn(array $m): bool =>
        $m['gospodarz'] === LIGA_NASZ_KLUB || $m['gosc'] === LIGA_NASZ_KLUB));

    $nadchodzace = array_values(array_filter($nasze, static fn(array $m): bool => !$m['wynik']));
    $rozegrane   = array_values(array_filter($nasze, static fn(array $m): bool => (bool) $m['wynik']));

    $dane = [
        'zrodlo'           => LIGA_URL,
        'liga'             => liga_nazwa($html),
        'klub'             => LIGA_NASZ_KLUB,
        'zaktualizowano'   => date('Y-m-d\TH:i:s'),
        'tabela'           => $tabela,
        'forma'            => liga_policz_forme($mecze),
        'ostatnia_kolejka' => liga_ostatnia_kolejka($mecze),
        'nadchodzace'      => array_slice($nadchodzace, 0, 8),
        'ostatni'          => $rozegrane ? end($rozegrane) : null,
        'wszystkie_mecze'  => $mecze,
    ];

    $tekst = liga_json($dane);

    if (!$dryRun) {
        $plik = __DIR__ . '/data/liga.json';

        if (!is_dir(dirname($plik))) {
            mkdir(dirname($plik), 0775, true);
        }

        if (file_put_contents($plik, $tekst . "\n") === false) {
            throw new RuntimeException('Brak prawa zapisu do ' . $plik);
        }
    }

    return [
        'tekst'       => $tekst,
        'tabela'      => count($tabela),
        'mecze'       => count($mecze),
        'nadchodzace' => count($nadchodzace),
    ];
}

/* Uruchamiamy się sami wyłącznie z wiersza poleceń (cron). Dołączony z panelu
   plik ma tylko udostępnić aktualizuj_lige(), a nie od razu ruszać po dane. */
if (PHP_SAPI === 'cli' && isset($argv[0]) && realpath($argv[0]) === realpath(__FILE__)) {
    try {
        $wynik = aktualizuj_lige(in_array('--dry-run', $argv, true));

        if (in_array('--dry-run', $argv, true)) {
            echo $wynik['tekst'], "\n";
        } else {
            printf(
                "Zapisano data/liga.json: %d zespołów w tabeli, %d meczów, %d nadchodzących.\n",
                $wynik['tabela'], $wynik['mecze'], $wynik['nadchodzace']
            );
        }
    } catch (Throwable $e) {
        fwrite(STDERR, 'Błąd: ' . $e->getMessage() . "\n");
        exit(1);
    }
}
